File: typescript/src/tries/design-add-and-search-words-data-structure.ts
/**
 * LeetCode 211. Design Add and Search Words Data Structure (Medium)
 * https://leetcode.com/problems/design-add-and-search-words-data-structure/
 *
 * Design a data structure supporting addWord(word) and search(word), where
 * search may contain '.' wildcards that match any single letter. All other
 * characters are lowercase English letters.
 *
 * Run just this file:   npx tsx src/tries/design-add-and-search-words-data-structure.ts
 * Run its tests:        npx vitest run src/tries/design-add-and-search-words-data-structure.test.ts
 */

export class WordDictionary {
  constructor() {
    // TODO: implement
  }

  addWord(word: string): void {
    throw new Error("Not implemented");
  }

  search(word: string): boolean {
    throw new Error("Not implemented");
  }
}

// Quick manual run / debugging playground — runs only when executed directly.
if (import.meta.url === `file://${process.argv[1]}`) {
  const wd = new WordDictionary();
  wd.addWord("bad");
  wd.addWord("dad");
  wd.addWord("mad");
  console.log(wd.search("pad")); // expected false
  console.log(wd.search("bad")); // expected true
  console.log(wd.search(".ad")); // expected true
  console.log(wd.search("b..")); // expected true
}

File: python/tries/design_add_and_search_words_data_structure.py
"""
LeetCode 211. Design Add and Search Words Data Structure (Medium)
https://leetcode.com/problems/design-add-and-search-words-data-structure/

Design a data structure supporting addWord(word) and search(word), where
search may contain '.' wildcards that match any single letter. All other
characters are lowercase English letters.

Run just this file:   python tries/design_add_and_search_words_data_structure.py
Run its tests:        pytest tries/design_add_and_search_words_data_structure.py -v
"""


class WordDictionary:
    def __init__(self):
        ...  # TODO: implement

    def addWord(self, word: str) -> None:
        ...  # TODO: implement

    def search(self, word: str) -> bool:
        ...  # TODO: implement


# ---------------------------- tests ----------------------------

def test_official_example():
    wd = WordDictionary()
    wd.addWord("bad")
    wd.addWord("dad")
    wd.addWord("mad")
    assert wd.search("pad") is False
    assert wd.search("bad") is True
    assert wd.search(".ad") is True
    assert wd.search("b..") is True


def test_wildcard_length_must_match():
    wd = WordDictionary()
    wd.addWord("bad")
    assert wd.search("b.") is False
    assert wd.search("b...") is False
    assert wd.search("...") is True


def test_all_wildcards_empty_dictionary():
    wd = WordDictionary()
    assert wd.search(".") is False
    wd.addWord("a")
    assert wd.search(".") is True


if __name__ == "__main__":
    # Debugging playground — set breakpoints here.
    wd = WordDictionary()
    wd.addWord("bad")
    wd.addWord("dad")
    wd.addWord("mad")
    print(wd.search("pad"))  # expected False
    print(wd.search("bad"))  # expected True
    print(wd.search(".ad"))  # expected True
    print(wd.search("b.."))  # expected True
